virtool/references/api.py

```diff
@@ -52,608 +68,723 @@
-    List all updates made to the reference.
+        elif data.remote_from:
+            try:
+                release = await virtool.github.get_release(
+                    self.request.app["config"],
+                    self.request.app["client"],
+                    data.remote_from,
+                    release_id=data.release_id,
+                )
 
-    """
-    db = req.app["db"]
-    ref_id = req.match_info["ref_id"]
+            except aiohttp.ClientConnectionError:
+                raise HTTPBadGateway(text="Could not reach GitHub")
 
-    if not await virtool.mongo.utils.id_exists(db.references, ref_id):
-        raise NotFound()
+            except GitHubError as err:
+                if "404" in str(err):
+                    raise HTTPBadGateway(
+                        text="Could not retrieve latest GitHub release"
+                    )
 
-    updates = await virtool.mongo.utils.get_one_field(db.references, "updates", ref_id)
+                raise
 
-    if updates is not None:
-        updates.reverse()
+            release = format_release(release)
 
-    return json_response(updates or [])
+            document = await create_remote(
+                db, settings, release, data.remote_from, user_id
+            )
 
+            context = {
+                "release": release,
+                "ref_id": document["_id"],
+                "created_at": document["created_at"],
+                "user_id": user_id,
+            }
 
-@routes.post("/refs/{ref_id}/updates")
-async def update(req):
-    app = req.app
-    db = app["db"]
+            task = await self.request.app["tasks"].add(
+                RemoteReferenceTask, context=context
+            )
 
-    ref_id = req.match_info["ref_id"]
-    user_id = req["client"].user_id
+            document["task"] = {"id": task["id"]}
+
+        else:
+            document = await create_document(
+                db,
+                settings,
+                data.name,
+                data.organism,
+                data.description,
+                data.data_type,
+                user_id=self.request["client"].user_id,
+            )
 
-    if not await virtool.mongo.utils.id_exists(db.references, ref_id):
-        raise NotFound()
+        await db.references.insert_one(document)
 
-    if not await virtool.references.db.check_right(req, ref_id, "modify"):
-        raise InsufficientRights()
+        headers = {"Location": f"/refs/{document['_id']}"}
 
-    release = await virtool.mongo.utils.get_one_field(db.references, "release", ref_id)
+        return json_response(
+            await attach_computed(db, document), headers=headers, status=201
+        )
 
-    if release is None:
-        raise HTTPBadRequest(text="Target release does not exist")
 
-    created_at = virtool.utils.timestamp()
+@routes.view("/refs/{ref_id}")
+@routes.jobs_api.get("/refs/{ref_id}")
+class ReferenceView(PydanticView):
+    async def get(self) -> Union[r200[ReferenceResponse], r404]:
```

Comment:
This should return `403` as well if the client doesn't have read permission on the reference.